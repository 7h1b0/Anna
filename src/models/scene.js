const mongoose = require('mongoose');
const Joi = require('joi');
const type = require('../utils/type');

const Schema = mongoose.Schema;

const scene = new Schema({
  name: { type: String, required: true },
  description: { type: String },
  actions: [
    {
      id: { type: String, required: true },
      name: { type: String },
      type: { type: String, required: true },
      body: { type: Object, required: true },
    },
  ],
});

scene.statics.validate = function validate(data, callback) {
  const pattern = {
    name: Joi.string()
      .trim()
      .min(3)
      .required(),
    description: Joi.string()
      .trim()
      .min(5)
      .optional(),
    actions: Joi.array()
      .min(1)
      .items(
        Joi.object().keys({
          id: Joi.any().required(),
          name: Joi.string().min(3),
          type: Joi.any()
            .valid(type.TYPES)
            .required(),
          body: Joi.object()
            .min(1)
            .optional(),
        }),
      ),
  };

  Joi.validate(data, pattern, callback);
};

module.exports = mongoose.model('Scene', scene);
