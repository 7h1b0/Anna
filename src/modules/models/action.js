const knex = require('../../knexClient');

const TABLE = 'actions';
const COLUMNS = ['name', 'type', 'body', { targetId: 'target_id' }];

module.exports = {
  TABLE,
  COLUMNS,

  findBySceneId(sceneId) {
    return knex(TABLE)
      .select(COLUMNS)
      .where('scene_id', sceneId)
      .then(entries => {
        return entries.reduce((acc, entry) => {
          try {
            return acc.concat(
              Object.assign({}, entry, { body: JSON.parse(entry.body) }),
            );
          } catch (e) {
            return acc;
          }
        }, []);
      });
  },
};
