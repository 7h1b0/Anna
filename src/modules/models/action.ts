import knex from '../../knexClient';
import TYPES from '../type';
import { ToggleDio, ToggleHueLight, AnnaAction } from '../actions';
import * as logger from '../logger';

export const TABLE = 'actions';

export interface Action {
  sceneId: string;
  name: string;
  targetId: string;
  readonly type: TYPES.HUE_LIGHT | TYPES.DIO;
  body: string;
}

export async function findBySceneId(sceneId: string): Promise<AnnaAction[]> {
  return knex(TABLE)
    .select(['type', 'body', 'targetId'])
    .where('sceneId', sceneId)
    .then((entries: Action[]) => {
      return entries.reduce((acc: AnnaAction[], entry: Action) => {
        try {
          const body = JSON.parse(entry.body);
          if (entry.type === TYPES.DIO) {
            return acc.concat([new ToggleDio(entry.sceneId, body)]);
          }
          return acc.concat(new ToggleHueLight(entry.sceneId, body));
        } catch (e) {
          logger.error(`Body of entry #${entry.name} is malformed`);
          return acc;
        }
      }, []);
    });
}
