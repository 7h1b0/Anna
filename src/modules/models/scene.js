const Ajv = require('ajv');
const knex = require('../../knexClient');
const sceneSchema = require('../schemas/scene');
const TABLE = 'scenes';
const COLUMNS = [
  { sceneId: 'scene_id' },
  'name',
  'description',
  { createdAt: 'created_at' },
  { updatedAt: 'updated_at' },
  { createdBy: 'created_by' },
];
const { TABLE: ACTION_TABLE, findBySceneId } = require('./action');

module.exports = {
  TABLE,
  COLUMNS,

  validate(data) {
    const ajv = new Ajv();
    return ajv.validate(sceneSchema, data);
  },

  findAll() {
    return knex(TABLE).select(COLUMNS);
  },

  findById(sceneId) {
    const fetchScene = knex(TABLE)
      .select(COLUMNS)
      .where('scene_id', sceneId);
    const fetchActions = findBySceneId(sceneId);

    return Promise.all([fetchScene, fetchActions]).then(([scene, actions]) => {
      if (scene.length > 0) {
        return Object.assign({}, scene[0], { actions });
      }
      return;
    });
  },

  findByIdAndUpdate({ sceneId, name, description, actions }) {
    return knex
      .transaction(trx => {
        return trx
          .update({ name, description, updated_at: new Date() })
          .where('scene_id', sceneId)
          .into(TABLE)
          .then(row => {
            if (row < 1) {
              throw new Error('No scene found');
            }
            return trx
              .del()
              .where('scene_id', sceneId)
              .into(ACTION_TABLE);
          })
          .then(() => {
            return Promise.all(
              actions.map(action => {
                const formatedAction = {
                  type: action.type,
                  name: action.name,
                  target_id: action.targetId,
                  scene_id: sceneId,
                  body: JSON.stringify(action.body),
                };
                return trx.insert(formatedAction).into(ACTION_TABLE);
              }),
            );
          });
      })
      .catch(err => {
        if (err.message === 'No scene found') {
          return 0;
        }
        throw err;
      });
  },

  save({ name, description, userId, actions }) {
    const date = new Date();
    return knex.transaction(trx => {
      return trx
        .insert({
          description,
          name,
          created_by: userId,
          updated_at: date,
          created_at: date,
        })
        .into(TABLE)
        .then(([sceneId]) => {
          return Promise.all(
            actions.map(action => {
              const formatedAction = {
                type: action.type,
                name: action.name,
                target_id: action.targetId,
                scene_id: sceneId,
                body: JSON.stringify(action.body),
              };
              return trx.insert(formatedAction).into(ACTION_TABLE);
            }),
          ).then(() => sceneId);
        });
    });
  },

  delete(sceneId) {
    return knex.transaction(trx => {
      const removeScene = trx
        .del()
        .where('scene_id', sceneId)
        .into(TABLE);

      const removeActions = trx
        .del()
        .where('scene_id', sceneId)
        .into(ACTION_TABLE);

      return Promise.all([removeScene, removeActions]);
    });
  },
};
