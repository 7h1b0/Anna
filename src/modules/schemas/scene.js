const { TYPES } = require('../type');

module.exports = {
  $id: 'scene',
  type: 'object',
  additionalProperties: false,
  properties: {
    sceneId: {
      type: 'integer',
    },
    name: {
      type: 'string',
      minLength: 3,
    },
    description: {
      type: 'string',
      minLength: 5,
    },
    actions: {
      type: 'array',
      minItems: 1,
      items: [
        {
          type: 'object',
          additionalProperties: false,
          properties: {
            targetId: {
              type: 'integer',
            },
            actionId: {
              type: 'integer',
            },
            name: {
              type: 'string',
              minLength: 3,
            },
            type: {
              type: 'string',
              enum: TYPES,
            },
            body: {
              type: 'object',
              minProperties: 1,
            },
          },
          required: ['targetId', 'name', 'type'],
        },
      ],
    },
  },
  required: ['name', 'actions'],
};
