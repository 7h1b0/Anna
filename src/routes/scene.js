const routes = require('express').Router();
const Scene = require('../modules/models/scene');
const { callScene } = require('../modules/actions');
const dispatch = require('../modules/dispatch');
const logger = require('../modules/logger');

routes
  .route('/api/scenes')
  .get((req, res) => {
    Scene.findAll()
      .then(scenes => res.json(scenes))
      .catch(err => res.status(500).send({ err }));
  })
  .post((req, res) => {
    const isValid = Scene.validate(req.body);
    if (!isValid) {
      return res.sendStatus(400);
    }

    const userId = res.locals.user.userId;
    Scene.save({ ...req.body, userId })
      .then(sceneId => res.status(201).json({ ...req.body, sceneId }))
      .catch(err => res.status(500).send({ err }));
  });

routes
  .route('/api/scenes/:id_scene([0-9]+)')
  .get((req, res) => {
    Scene.findById(req.params.id_scene)
      .then(scene => {
        if (!scene) {
          res.sendStatus(404);
        } else {
          res.json(scene);
        }
      })
      .catch(err => res.status(500).send({ err }));
  })
  .patch((req, res) => {
    const isValid = Scene.validate(req.body);
    if (!isValid) {
      return res.sendStatus(400);
    }
    Scene.findByIdAndUpdate({ ...req.body, sceneId: req.params.id_scene })
      .then(rowsAffected => {
        if (!!rowsAffected) {
          res.sendStatus(204);
        } else {
          res.sendStatus(404);
        }
      })
      .catch(err => res.status(500).send({ err }));
  })
  .delete((req, res) => {
    Scene.delete(req.params.id_scene)
      .then(([removedScene]) => {
        if (removedScene < 1) {
          res.sendStatus(404);
        } else {
          res.sendStatus(204);
        }
      })
      .catch(err => {
        res.status(500).send({ err });
      });
  });

routes.get('/api/scenes/:id_scene([0-9]+)/action', (req, res) => {
  dispatch(callScene(req.params.id_scene))
    .then(() => res.end())
    .catch(err => {
      logger.error(err);
      res.status(500).send({ err });
    });
});

module.exports = routes;
