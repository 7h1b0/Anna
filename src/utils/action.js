const Dio = require('../models/dio');
const Scene = require('../models/scene');
const type = require('./type');
const hueService = require('../services/hueService');

module.exports = function dispatch(actions) {
  actions.forEach(action => {
    switch (action.type) {
      case type.HUE_GROUP:
        hueService.setGroupState(action.id, action.body);
        break;

      case type.HUE_LIGHT:
        hueService.setLightState(action.id, action.body);
        break;

      case type.DIO:
        Dio.updateState(action.id, action.body.on);
        break;

      case type.SCENE:
        Scene.findById(action.id).then(scene => {
          if (scene) dispatch(scene.actions);
        });
        break;

      default:
        break;
    }
  });
};
