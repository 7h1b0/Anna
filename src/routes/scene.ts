import { Router } from 'express';
import * as Scene from 'modules/models/scene';
import { callScene } from 'modules/actions';
import dispatch from 'modules/dispatch';
import * as logger from 'modules/logger';

const routes = Router();
routes
  .route('/api/scenes')
  .get((req, res) => {
    Scene.findAll()
      .then(scenes => res.json(scenes))
      .catch(err => res.status(500).send({ err }));
  })
  .post((req, res) => {
    const isValid = Scene.validate(req.body);
    if (!isValid) {
      res.sendStatus(400);
    } else {
      const userId = res.locals.user.userId;
      Scene.save({ ...req.body, userId })
        .then(sceneId => res.status(201).json({ sceneId }))
        .catch(err => res.status(500).send({ err }));
    }
  });

routes
  .route('/api/scenes/:sceneId([a-fA-F0-9-]{36})')
  .get((req, res) => {
    Scene.findById(req.params.sceneId)
      .then(scene => {
        if (!scene) {
          res.sendStatus(404);
        } else {
          res.json(scene);
        }
      })
      .catch(err => {
        res.status(500).send({ err });
      });
  })
  .patch((req, res) => {
    const isValid = Scene.validate(req.body);
    if (!isValid) {
      res.sendStatus(400);
    } else {
      Scene.findByIdAndUpdate({ ...req.body, sceneId: req.params.sceneId })
        .then(rowsAffected => {
          if (!!rowsAffected) {
            res.sendStatus(204);
          } else {
            res.sendStatus(404);
          }
        })
        .catch(err => res.status(500).send({ err }));
    }
  })
  .delete((req, res) => {
    Scene.remove(req.params.sceneId)
      .then(([removedScene]) => {
        if (removedScene < 1) {
          res.sendStatus(404);
        } else {
          res.sendStatus(204);
        }
      })
      .catch(err => {
        res.status(500).send({ err });
      });
  });

routes.get('/api/scenes/:sceneId([a-fA-F0-9-]{36})/action', (req, res) => {
  dispatch(callScene(req.params.sceneId))
    .then(() => res.end())
    .catch(err => {
      logger.error(err);
      res.status(500).send({ err });
    });
});

export default routes;
