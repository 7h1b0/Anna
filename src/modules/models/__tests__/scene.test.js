import MockDate from 'mockdate';
import knex from '../../../knexClient';
import * as Scene from '../scene';
import * as Action from '../action';
const initScenes = [
  {
    scene_id: 1,
    description: 'this is a test',
    name: 'scene_1',
    created_by: 1,
    created_at: new Date('2018-01-01'),
    updated_at: new Date('2018-01-02'),
  },
  {
    scene_id: 2,
    description: 'this is a second test',
    name: 'scene_2',
    created_by: 1,
    created_at: new Date('2018-01-01'),
    updated_at: new Date('2018-01-02'),
  },
];

const initActions = [
  {
    action_id: 1,
    scene_id: 1,
    type: 'DIO',
    name: 'action turn on',
    target_id: 1,
    body: JSON.stringify({ on: true }),
  },
  {
    action_id: 2,
    scene_id: 1,
    type: 'DIO',
    name: 'action turn off',
    target_id: 2,
    body: JSON.stringify({ on: false }),
  },
  {
    action_id: 3,
    scene_id: 2,
    type: 'SCENE',
    name: 'call scene',
    target_id: 2,
    body: null,
  },
];

describe('Scene', () => {
  beforeAll(async () => {
    await Promise.all([
      knex(Scene.TABLE).truncate(),
      knex(Action.TABLE).truncate(),
    ]);
  });

  beforeEach(async () => {
    await Promise.all([
      knex(Scene.TABLE).insert(initScenes),
      knex(Action.TABLE).insert(initActions),
    ]);
  });

  afterEach(async () => {
    MockDate.reset();
    await Promise.all([
      knex(Scene.TABLE).truncate(),
      knex(Action.TABLE).truncate(),
    ]);
  });

  afterAll(async () => {
    await knex.destroy();
  });

  describe('findAll', () => {
    it('should return all scene', async () => {
      const result = await Scene.findAll();
      expect(result).toMatchSnapshot();
    });
  });

  describe('delete', () => {
    it('should delete a scene', async () => {
      await Scene.remove(1);
      const scenes = await knex(Scene.TABLE)
        .select('*')
        .where('scene_id', 1);
      const actions = await knex(Action.TABLE)
        .select('*')
        .where('scene_id', 1);
      expect(scenes).toHaveLength(0);
      expect(actions).toHaveLength(0);
    });

    it('should not delete a scene', async () => {
      await Scene.remove(-1);
      const scenes = await knex(Scene.TABLE).select('*');
      const actions = await knex(Action.TABLE).select('*');
      expect(scenes).toEqual(initScenes);
      expect(actions).toEqual(initActions);
    });
  });

  describe('save', () => {
    it('should save a new scene', async () => {
      MockDate.set('2018-05-05');
      const scene = {
        description: 'save test',
        name: 'scene_3',
        userId: 1,
        actions: [
          {
            targetId: 1,
            name: 'action one',
            type: 'DIO',
            body: {
              on: true,
            },
          },
          {
            targetId: 1,
            name: 'action two',
            type: 'HUE_LIGHT',
            body: {
              on: true,
              bri: 255,
            },
          },
        ],
      };

      const id = await Scene.save(scene);
      expect(id).toBe(3);

      const scenes = await knex(Scene.TABLE)
        .select('*')
        .where('scene_id', 3);
      const actions = await knex(Action.TABLE)
        .select('*')
        .where('scene_id', 3);
      expect(scenes).toMatchSnapshot();
      expect(actions).toMatchSnapshot();
    });
  });

  describe('findByIdAndUpdate', () => {
    it('should update a scene', async () => {
      MockDate.set('2018-05-05');
      const updatedScene = {
        sceneId: 2,
        description: 'this is an updated second test',
        name: 'scene_2',
        actions: [
          {
            type: 'DIO',
            name: 'action turn off',
            targetId: 2,
            body: { on: false },
          },
          {
            type: 'DIO',
            name: 'action turn off',
            targetId: 4,
            body: { on: true },
          },
        ],
      };

      await Scene.findByIdAndUpdate(updatedScene);

      const scenes = await knex(Scene.TABLE)
        .select('*')
        .where('scene_id', 2);
      const actions = await knex(Action.TABLE)
        .select('*')
        .where('scene_id', 2);
      expect(scenes).toMatchSnapshot();
      expect(actions).toMatchSnapshot();
    });

    it('should not update a scene if sceneId is unknow', async () => {
      const updatedScene = {
        sceneId: 6,
        description: 'this is an updated second test',
        name: 'scene_2',
        actions: [
          {
            type: 'DIO',
            name: 'action turn off',
            targetId: 2,
            body: { on: false },
          },
          {
            type: 'DIO',
            name: 'action turn off',
            targetId: 4,
            body: { on: true },
          },
        ],
      };

      const res = await Scene.findByIdAndUpdate(updatedScene);
      expect(res).toBe(0);

      const scenes = await knex(Scene.TABLE)
        .select('*')
        .where('scene_id', 6);
      expect(scenes).toHaveLength(0);
      const actions = await knex(Action.TABLE)
        .select('*')
        .where('scene_id', 6);
      expect(actions).toHaveLength(0);
    });
  });

  describe('validate', () => {
    it('should return true when a scene is valid', () => {
      const scene = {
        name: 'test',
        description: 'my scene test',
        actions: [
          {
            targetId: 1,
            type: 'DIO',
            body: {
              on: true,
            },
          },
        ],
      };

      expect(Scene.validate(scene)).toBeTruthy();
    });

    it('should return false when no actions is provided', () => {
      const scene = {
        name: 'test',
        description: 'my scene test',
      };

      expect(Scene.validate(scene)).toBeFalsy();
    });

    it('should return false when an actions is missing props', () => {
      const scene = {
        name: 'test',
        description: 'my scene test',
        actions: [
          {
            targetId: 1,
            name: 'test',
            body: {
              on: true,
            },
          },
        ],
      };

      expect(Scene.validate(scene)).toBeFalsy();
    });

    it('should return true when no description is provided', () => {
      const scene = {
        name: 'test',
        actions: [
          {
            targetId: 1,
            name: 'test',
            type: 'DIO',
            body: {
              on: true,
            },
          },
        ],
      };

      expect(Scene.validate(scene)).toBeTruthy();
    });

    it('should return false when the type of an actions is unknow', () => {
      const scene = {
        name: 'test',
        actions: [
          {
            targetId: 1,
            name: 'test',
            type: 'TEST',
            body: {
              on: true,
            },
          },
        ],
      };

      expect(Scene.validate(scene)).toBeFalsy();
    });
  });
});
