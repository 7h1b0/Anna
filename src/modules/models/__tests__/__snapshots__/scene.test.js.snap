// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scene delete should delete a scene 1`] = `
Array [
  Object {
    "description": "this is a second test",
    "name": "scene_2",
    "scene_id": 2,
  },
]
`;

exports[`Scene delete should delete a scene 2`] = `
Array [
  Object {
    "action_id": 3,
    "body": null,
    "name": "call scene",
    "scene_id": 2,
    "target_id": 2,
    "type": "SCENE",
  },
]
`;
